src/irr/asset/CGraphicsPipelineLoaderMTL.cpp

```diff
@@ -0,0 +1,467 @@
+        auto bindings = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUDescriptorSetLayout::SBinding>>(textureCnt);
+
+        ICPUDescriptorSetLayout::SBinding bnd;
+        bnd.count = 1u;
+        bnd.stageFlags = ICPUSpecializedShader::ESS_FRAGMENT;
+        bnd.type = EDT_COMBINED_IMAGE_SAMPLER;
+        std::fill(bindings->begin(), bindings->end(), bnd);
+
+        ICPUSampler::SParams smplParams;
+        smplParams.TextureWrapU = smplParams.TextureWrapV = smplParams.TextureWrapW = ICPUSampler::ETC_REPEAT;
+        smplParams.BorderColor = ICPUSampler::ETBC_FLOAT_TRANSPARENT_BLACK;
+        smplParams.MinFilter = smplParams.MaxFilter = ICPUSampler::ETF_LINEAR;
+        smplParams.MipmapMode = ICPUSampler::ESMM_LINEAR;
+        smplParams.AnisotropicFilter = 0u;
+        smplParams.CompareEnable = 0u;
+        smplParams.LodBias = 0.f;
+        smplParams.MinLod = -1000.f;
+        smplParams.MaxLod = 1000.f;
+
+        core::smart_refctd_ptr<ICPUSampler> samplers[2];
+        samplers[0] = core::make_smart_refctd_ptr<ICPUSampler>(smplParams);
+        smplParams.TextureWrapU = smplParams.TextureWrapV = smplParams.TextureWrapW = ICPUSampler::ETC_CLAMP_TO_BORDER;
+        samplers[1] = core::make_smart_refctd_ptr<ICPUSampler>(smplParams);
+        for (uint32_t i = 0u, j = 0u; i <= CMTLPipelineMetadata::SMtl::EMP_REFL_POSX; ++i)
+        {
+            if (!_mtl.maps[i].empty())
+            {
+                (*bindings)[j].binding = j;
+
+                const uint32_t clamp = (_mtl.clamp >> i) & 1u;
+                (*bindings)[j].samplers = samplers + clamp;
+
+                ++j;
+            }
+        }
+
+        auto dsLayout = core::make_smart_refctd_ptr<ICPUDescriptorSetLayout>(bindings->begin(), bindings->end());
+        //ds with textures for material goes to set=3
+        auto layout = core::make_smart_refctd_ptr<ICPUPipelineLayout>(nullptr, nullptr, nullptr, nullptr, nullptr, std::move(dsLayout));
+
+        return layout;
+    }
+}
+
+SAssetBundle CGraphicsPipelineLoaderMTL::loadAsset(io::IReadFile* _file, const IAssetLoader::SAssetLoadParams& _params, IAssetLoader::IAssetLoaderOverride* _override, uint32_t _hierarchyLevel)
+{
+    auto materials = readMaterials(_file);
+
+    core::vector<core::smart_refctd_ptr<ICPURenderpassIndependentPipeline>> pipelines(materials.size());
+    SVertexInputParams vtxParams;
+    SBlendParams blendParams;
+    SPrimitiveAssemblyParams primParams;
+    SRasterizationParams rasterParams;
+    for (size_t i = 0ull; i < pipelines.size(); ++i)
+    {
+        auto layout = makePipelineLayoutFromMtl(materials[i]);
+        pipelines[i] = core::make_smart_refctd_ptr<ICPURenderpassIndependentPipeline>(nullptr, std::move(layout), nullptr, nullptr, vtxParams, blendParams, primParams, rasterParams);
+        m_assetMgr->setAssetMetadata(pipelines[i].get(), core::make_smart_refctd_ptr<CMTLPipelineMetadata>(std::move(materials[i])));
+    }
```

Comment:
actually we should go the **UBERSHADER** route here, which is one shader with all the material models, where you only switch between them using a uniform or a push constant (whichever one you like).

The metadata returned will help the user build the "expected" descriptor set contents.

Reply:
The idea should be that your loader for MTL (and to be honest any other loader, except for .baw) adds big ubershaders to the asset manager in the constructor of the loader (when it gets added).

Then we only really compile once when loading the models of a particular format, AND no have to deal with "fake" shader swaps on the API side.

You should cache your pipeline layouts too, since you'll probably be returning the same one over and over again (unless you choose to differentiate for different texture set-ups).

Reply:
in reality the pipelines themselves could also be precomputed and during the load, cloned appropriately (they only need to be unique objects because the metadata changes, but internals will be identical)

It will be important to implement a correct `operator==` for `core::unordered_set` for pipelines, shaders and layouts that can reduce duplicates in the default cpu to gpu object converter.